providers/base/bin/tgpio.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+import argparse
+import subprocess
+import shlex
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        prog="TGPIO Testing Tool",
+        description="This is a tool to help you perform the TGPIO testing",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+    parser.add_argument(
+        "-r",
+        "--run",
+        type=str,
+        choices=["server", "receive"],
+        help="The ptp device node to test",
+        required=True,
+    )
+    parser.add_argument(
+        "-d",
+        "--device",
+        type=str,
+        help="The ptp device node to test",
+        required=True,
+    )
+    parser.add_argument(
+        "-p",
+        "--pin",
+        type=str,
+        help="The pin number to test",
+        required=True,
+    )
+    parser.add_argument(
+        "-t",
+        "--timeout",
+        type=int,
+        help="The timeout in seconds",
+        default=10,
+    )
+    return parser.parse_args()
+
+
+def server_mode(device: str, pin: str):
+    cmd = "testptp -d {} -i {} -L {},2 -p 1000000000".format(device, pin, pin)
+    subprocess.run(shlex.split(cmd))
+
+
+def receive_timestamp(device: str, pin: str, timeout: int = 10):
+    cmd = "testptp -d {} -i {} -L {},1 -e 100 -o {} -E".format(
+        device, pin, pin, timeout
+    )
+    result = subprocess.run(
+        shlex.split(cmd),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+    print(result.stdout)
+    if result.returncode:
+        raise SystemExit(
+            "[ERROR] Failed to receive timestamp: {}".format(result.stderr)
+        )
+
+    cnt = 0
+    prev_event_time = -1
+    for line in result.stdout.splitlines():
+        if "Event time time" not in line:
+            continue
+        cnt += 1
+        event_time = int(line.split(":")[-1].split(",")[0].strip())
+        if prev_event_time == -1:
+            prev_event_time = event_time
+            continue
+        delta = event_time - prev_event_time
+        if delta != 1:
+            raise SystemExit("[ERROR] Time Delta is not 1")
+        prev_event_time = event_time
+    if cnt < timeout - 1 or timeout + 1 < cnt:
+        raise SystemExit("[ERROR] The number for Event time time is incorrect")
+    print("[PASS] Time Delta is 1")
```

Comment:
I feel like this is slightly more readable, see if you like it: (double check if this is correct, I'm writing on github)
```suggestion
    try:
        output = subprocess.check_output(
            shlex.split(cmd)
        )
    except subprocess.CalledProcessError as e:
        raise SystemExit(
            "[ERROR] Failed to receive timestamps: {}".format(str(e))
        )
    
    event_times =[
        int(line.split(":")[-1].split(",")[0].strip())
        for line in output.splitlines()
        if "Event time time" in line
    ]

    cnt = len(event_times)
    deltas = deltas = [
        y - x for (x,y) in zip(iter(event_times), iter(event_times[1:]))
    ]
    if any(delta != 1 in deltas):
        raise SystemExit("[ERROR] Some events didn't take exactly 1\nTimestamps: {}\nDeltas:{}".format(str(event_times), str(deltas))
    if cnt < timeout - 1 or timeout + 1 < cnt:
        raise SystemExit("[ERROR] The number for Event time time is incorrect")
    print("[PASS] Time Delta is 1")
```

Questions: 
- What comes after the comma in `Event time time: TIME,...` (the lines we are parsing)
- Why check if `timeout + 1 < cnt`? Shouldn't `timeout - 2 < cnt < timeout`? (as if we are starting at X, we may loose X timestamp and X + timeout timestamp if we don't start at int(X), while if we start at int(X) we always get at least tiemout - 1 cnt)

Reply:
1. It is the nano second, and currently we only care about the second.
2. Firstly, the TGPIO timestamps may not be perfectly one second. Secondly, if we are starting at X, we may occasionally have the X time stamp, and have the X+10 at the end. That's why it may be 11 timestamps. You could refer to the result [tgpio/ptp-time-check-/dev/ptp1-0](https://certification.canonical.com/hardware/202311-32338/submission/368615/test/210867/result/40832596/).

Reply:
By the way, TGPIO may not be able to provide precise time, it's responsibility is to sync up the time between 2 systems. That's why I said the timestamps provided by TGPIO may not be exactly one second.